### PR TITLE
dataset splitting

### PR DESCRIPTION
**Pull request by mkos**
### Summary

Provides a helper function that allows splitting the dataset to training and test sets.

### Other Information

Supports sample-axis-wise and time-axis-wise directions, lazy loading etc.

### CODE REVIEW

**Review by hrzn — COMMENTED** [review]
Looks good @mkos 
Just have a few comments but nothing major.

File: darts/utils/model_selection.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        self.type = type
+
+        if not data:
+            raise AttributeError('The `data` parameter cannot be empty.')
+        self.data = data
+
+        self.test_size = test_size
+        self.axis = axis
+
+        if axis == 1 and vertical_split_type == 'model-aware' and (horizon == 0 or input_size == 0):
```
[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py:27]: 👍🏻 for constants. Enums are not so convenient for input parameters, so didn't go there.

File: darts/utils/model_selection.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                return train_set, test_set
+            else:
+                return list(train_set), list(test_set)
+
+
+def train_test_split(
+        data: Union[TimeSeries, Sequence[TimeSeries]],
+        test_size: Optional[Union[float, int]] = 0.25,
+        axis: Optional[int] = 0,
+        input_size: Optional[int] = 0,
```
[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py:147]: Done. also updated docstrings a bit.

File: darts/utils/model_selection.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+    def __getitem__(self, i: int) -> TimeSeries:
+        if self.axis == 0:
+            split_index = self._get_horizontal_split_index()
+            if self.type == 'train':
+                if i >= split_index:
+                    raise IndexError('Exceeded the size of the sequence.')
+                return self.data[i]
+            else:
+                if i + split_index > len(self.data):
```
[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py:90]: It can't be `>=` since we are comparing 0-based indicies to length of the array. To satisfy `>=` it would have to be `i + split_index >= len(self.data) - 1`, because `len(self.data) - 1` gives the last index.

[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py:90]: there are tests that test it, btw: `test_horiz_split_overindexing_test_set()`, `test_horiz_split_last_index_test_set()`

File: darts/utils/model_selection.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        input_size
+            size of the input [default: 0]
+
+        horizon
+            forecast horizon [default: 0]
+
+        vertical_split_type
+            can be either ``simple``, where the exact number from test size will be deducted from timeseries for test set and
+            remaining will go to training set; or ``model-aware``, where you have to provide ``input_size`` and ``horizon``
+            as well. Note, that second option is more efficient timestep-wise, since training and test sets will be
```
[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py]: For clarity, I didn't add it in the parameter description but in the main comment.

File: darts/utils/model_selection.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+SIMPLE = 'simple'
+
+
+class SplitTimeSeriesSequence(Sequence):
+    """
+    This class is primarily meant to be instantiated from ``train_test_split()`` function.
+    """
+    def __init__(self,
+                 type: str,
+                 data: Union[TimeSeries, Sequence[TimeSeries]],
```
[the remark this answers is not in the document]
Comment (hrzn) [reply darts/utils/model_selection.py:14]: Good catch

File: darts/utils/model_selection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hrzn) [reply darts/utils/model_selection.py:154]: I still see a default value of `0`, or is it me?

File: darts/utils/model_selection.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hrzn) [reply darts/utils/model_selection.py:97]: Yes and I guess the tests are not catching it because in the `=` case line 99 below will also anyway return an `IndexError`, which is what your test expects.

File: darts/utils/model_selection.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+                test_size = self.test_size
+
+            test_start_index = ts_length - test_size
+            train_end_index = test_start_index
+
+            if test_start_index < 0:
+                raise AttributeError("`test_size` is bigger then timeseries length")
+
+        else: # model-aware split
+            train_end_index = ts_length - self.horizon
```
Comment (hrzn) [inline darts/utils/model_selection.py:69]: This means that models might be trained on some parts of the test set. There can even be cases where some whole input/output sub-sequences in the test set also appear in the training set. I guess this is not what we want, right? Because that wouldn't be a "true" test set. 

I would have expected something like `train_end_index = test_start_index + input_size` because you can have one "input" worth of overlap but not one "input + output" of overlap without leakage (unless I'm missing something).

File: darts/utils/model_selection.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            else:
+                test_size = self.test_size
+
+            if train_end_index < 0:
+                train_end_index = 0
+
+            if train_end_index < self.input_size:
+                raise AttributeError("Training timeseries is of 0 size")
+
+            test_start_index = ts_length - self.horizon - self.input_size - test_size - 1
```
[the remark this answers is not in the document]
Comment (hrzn) [reply darts/utils/model_selection.py:82]: I think -1 works

[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py:82]: No, Samuele is right. I have attached the example in the code comments

[the remark this answers is not in the document]
Comment (hrzn) [reply darts/utils/model_selection.py]: Thanks

**Review by hrzn — APPROVED** [review]
👍

File: darts/utils/model_selection.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+            train_end_index = test_start_index
+
+            if test_start_index < 0:
+                raise AttributeError("`test_size` is bigger then timeseries length")
+
+        else: # model-aware split
+
+            # an example on how these are calculated:
+            # test_size = 4
+            # input_size = 5 (1)
```
Comment (piaz97) [inline darts/utils/model_selection.py:72]:
```suggestion
            # input_size = 5 (-)
```

File: darts/utils/model_selection.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+            # len(data) = 16 (*)
+
+            # 0         5       9           15 <- index
+            # * * * * * * * * * * * * * * * *
+            #                 - - - - - + + +   < test sample 1
+            #               - - - - - + + +     < test sample 2
+            #             - - - - - + + +       < test sample 3
+            #           - - - - - + + +         < test sample 4
+            #           ^       ^
+            #           |       train_end_index = 9
```
Comment (piaz97) [inline darts/utils/model_selection.py:83]: `train_end_index = 16 - 3 - 4 + 1 = 10` (currently 9). So also the `^` should be moved right, under the first `+`, and the index on top too (Which should make sense since afterward you select `[:train_end_index]`, so  `train_end_index` should be the first `+` indeed)

File: darts/utils/model_selection.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+    When splitting over the time axis, splitter tries to greedy satisfy the requested test set size, i.e. when one of
+    the timeseries in sequence is too small, all samples will go to the test set and the exception will be raised.
+
+
+    Time axis split with ``model-aware`` split type enabled tries to reclaim as much datapoints for training as
+    possible by partially overlapping test set with training set. This is possible because only forecasted part of
+    test set cannot be used for training. The formula to calculate last available timestep for training set is
+    following:
+
+        train end index = timeseries length - horizon
```
Comment (piaz97) [inline darts/utils/model_selection.py:191]:
```suggestion
        train end index = timeseries length - horizon - test_size
```
According to line 91, train_end_index = ts_length - self.horizon - test_size + 1. Since the comment mention the last _available_ index, then the +1 should not be necessary in this formula

File: darts/utils/model_selection.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py:114]: Oooomph. I thought that `assertRaise()` parameter `msg=` checks whether message from the exception is correct. Turns out that you have to use `assertRaiseRegex()` to do that.
